pkg/metrics/otelmetrics/otelCounter.go

```diff
@@ -0,0 +1,37 @@
+package otelmetrics
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+type otelCounter struct {
+	counter      metric.Int64Counter
+	attributeSet attribute.Set
+}
+
+func newOtelCounter(name string, tags map[string]string) *otelCounter {
+	meter := otel.Meter("jaeger-V2")
+	counter, err := meter.Int64Counter(name)
+	if err != nil {
+		panic(err)
+	}
+
+	attributes := make([]attribute.KeyValue, 0, len(tags))
+	for k, v := range tags {
+		attributes = append(attributes, attribute.String(k, v))
+	}
+	attributeSet := attribute.NewSet(attributes...)
+
+	return &otelCounter{
+		counter:      counter,
+		attributeSet: attributeSet,
+	}
+}
+
+func (c *otelCounter) Inc(value int64) {
+	c.counter.Add(context.Background(), value, metric.WithAttributeSet(c.attributeSet))
```

Comment:
We said we don't want allocations in this function.

Reply:
Ok I just want to run through my current understanding to you, Counter is an interface, I have created a otelCounter struct that implements the Inc function so it satisfies the definition of Counter Interface.
The Inc function handles incrementing and hence the counter.add function should be called in it.
Hence, otelCounter will be an implentation of the Counter metric api.
Is my understanding correct?

Reply:
That's all fine. But when you call Add() it has to be very quick. Instead you are creating extra objects before calling Add, objects that don't change once the counter struct is created. So pre-create them. And write a benchmark to compare with other implementations, it's very easy to do in Go.

Reply:
Ohh okay, I understand that now. Working on it right now.

Reply:
![Screenshot 2024-06-21 201714](https://github.com/jaegertracing/jaeger/assets/103821431/175bade6-75eb-40d1-9a28-36b15c251eb5)
I was actually following this section, where they called the WithAtrributes() in Add function. Thanks for the clarification.

Reply:
Sure, that's a naive example, but think about what it means. A counter is ultimately a location in memory with a value that you atomically increment. If you don't have a direct reference to an object that knows that memory location, you need to do some preprocessing of the arguments, like taking the attributes, computing some hash from them, and looking up the memory location in some hashtable. That's like orders of magnitude more CPU cycles than our existing implementation: counter.Inc -> promCounter.Int -> memLoc+=1.

I am actually surprised that there is no API in OTEL to get to a "fully resolved" counter which only has Inc(n) without additional attributes. E.g. something like `c1 := c2.WithAttributes()` - we could call this once at construction and never worry about passing arguments again.

Reply:
but again, theoretical debates about performance are much less useful than a straight benchmark. Maybe a benchmark would show that the above example is only 10% slower than our existing API, in which case we wouldn't need to bother.

Reply:
Ah okay, understand it now why it was an expensive operation. So, I have to draft a benchmark first, compare the existing API with the OTEL implementation  and based on the results you will decide if we should continue with it?
But, I had a question if the OTEL implementation of initializing the counter while mandatorily passing the attributes in each call is definitely going to be slower. Why even bother switching to the OTEL implementation of metrics?

Reply:
Well, first, politically it would be pretty weird to say that "Jaeger is fully on board with OTEL, but we don't think users should use OTEL SDK before performance is bad". And we don't know it's bad, that's why we need a benchmark.

But second, the OTEL Collector itself is definitely going to use OTEL for metrics, and we're reusing a whole bunch of collector components, so it doesn't really make sense to have a split brain and have some metrics being emitted by OTEL SDK while others by Prometheus SDK directly as in jaeger-v1.

Reply:
Ah yes, makes sense.

Reply:
Hi, I have created the benchmark file to compare the different implementations and have attached the result in the screenshot.
![Screenshot 2024-06-22 231030](https://github.com/jaegertracing/jaeger/assets/103821431/d80dd182-7d05-4bb5-87f5-dd7c079af454)

Reply:
Well, I think there is some kind of mistake in my implementation because OTEL implementation is a staggering 745% percentage slower than the prometheus one? 😥😥